crates/sui-indexer/src/store/indexer_store_v2.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_trait::async_trait;
+use move_binary_format::CompiledModule;
+use move_bytecode_utils::module_cache::GetModule;
+use std::sync::Arc;
+
+use sui_types::base_types::{ObjectID, SequenceNumber};
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use sui_types::object::ObjectRead;
+
+use crate::errors::IndexerError;
+use crate::handlers::{EpochToCommit, TransactionObjectChangesToCommit};
+use crate::metrics::IndexerMetrics;
+
+use crate::types_v2::{
+    IndexedCheckpoint, IndexedEvent, IndexedPackage, IndexedTransaction, TxIndex,
+};
+
+#[async_trait]
+pub trait IndexerStoreV2 {
```

Comment:
Are we sure having a trait like this is valuable? As in do we expect to abstract away the Indexer store and implement this trait for multiple types or do we expect to only have it implemented for the one main store (Postgres)? I ask because this may not provide as much value as we expect it to, especially for the graphql implementation side of things. Graphql allows for the request to specify which fields are wanted, and as such i expect that for the reading side of things that, that implementation will need a direct handle to a DB connection in order to only grab the data that is wanted from a particular row, vs the whole row.

If this is only intended to be used for the writing side of things, then i can see a bit more value as when we do writes we do whole rows at a time, or a batch of rows.

Thoughts?

Reply:
I haven't put too much thinking into this so my thought may not be throughout enough:

1. For write path, we know what interfaces we are looking for, mostly. This trait would do the work.
2. For read path, I agree what we have in `trait IndexerStore` will not work for graphql. But a) for different storage, the read implementation would differ too and b) before we sunset jsonrpc fully, we still need to wire up those queries. So a question is to answer is should we use a shared layer between jsonrpc query & graphql query, or a standalone thing for jsonrpc, which could be a separate trait. I'd prefer the former but I don't mind the latter if we have strong reasons.